pkg/proxy/handler.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apiserver/pkg/endpoints/filters"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	kubernetesscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/klog/v2"
+
+	tenancyhelper "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1/helper"
+	kcpauthorization "github.com/kcp-dev/kcp/pkg/authorization"
+	"github.com/kcp-dev/kcp/pkg/proxy/index"
+)
+
+func shardHandler(index index.Index, proxy http.Handler) http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		var comps = strings.SplitN(strings.TrimLeft(req.URL.Path, "/"), "/", 3)
+		if len(comps) != 3 || comps[0] != "clusters" {
+			http.NotFound(w, req)
+			return
+		}
+
+		ctx := req.Context()
+		attributes, err := filters.GetAuthorizerAttributes(ctx)
+		if err != nil {
+			responsewriters.InternalError(w, req, err)
+			return
+		}
+
+		clusterName := logicalcluster.New(comps[1])
+		if !tenancyhelper.IsValidCluster(clusterName) {
+			// this includes wildcards
+			klog.V(4).Infof("Invalid cluster name %q", req.URL.Path)
+			responsewriters.Forbidden(req.Context(), attributes, w, req, kcpauthorization.WorkspaceAcccessNotPermittedReason, kubernetesscheme.Codecs)
+			return
+		}
+
+		shardURLString, found := index.Lookup(clusterName)
+		if !found {
+			klog.V(4).Infof("Unknown cluster %q", clusterName)
+			responsewriters.Forbidden(req.Context(), attributes, w, req, kcpauthorization.WorkspaceAcccessNotPermittedReason, kubernetesscheme.Codecs)
+			return
+		}
+		shardURL, err := url.Parse(shardURLString)
```

Comment:
Why not store the parsed URL in the index?

Reply:
a `string` will take less space than a struct

Reply:
and I assume that it is valid anyway because it comes from validated API objects. So it will not change anything correctness-wise.

Reply:
@p0lyn0mial what do you mean with less space? If the first thing you do when you remove it from the index is parse it, having one parsed value in the index might even be _less_ space with large concurrent access, no? :)

@stts you do not assume it to be valid, there's an error check right here ... lol

Reply:
(small nit, can ignore)

Reply:
@stevekuznetsov stroing `url.URL` (struct) vs `string` (type). I think we would need at least `2` bytes for every string (pointer, length + actual data). The `url.URL` struct has a lot of fields.